atom/src/chunk.rs

```diff
@@ -0,0 +1,171 @@
+
+/// An atom containing memory of undefined type.
+///
+/// [See also the module documentation.](index.html)
+pub struct Chunk;
+
+unsafe impl UriBound for Chunk {
+    const URI: &'static [u8] = sys::LV2_ATOM__Chunk;
+}
+
+impl URIDBound for Chunk {
+    type CacheType = AtomURIDCache;
+
+    fn urid(urids: &AtomURIDCache) -> URID<Self> {
+        urids.chunk
+    }
+}
+
+impl<'a, 'b> Atom<'a, 'b> for Chunk
+where
+    'a: 'b,
+{
+    type ReadParameter = ();
+    type ReadHandle = &'a [u8];
+    type WriteParameter = ();
+    type WriteHandle = ByteWriter<'a, 'b>;
+
+    fn read(space: Space<'a>, _: ()) -> Option<&'a [u8]> {
+        space.data()
+    }
+
+    fn init(frame: FramedMutSpace<'a, 'b>, _: ()) -> Option<ByteWriter<'a, 'b>> {
+        Some(ByteWriter::new(frame))
+    }
+}
+
+/// A blob writer.
+///
+/// This struct is able to copy data into an atom. It basically wraps a `FramedMutSpace`.
+pub struct ByteWriter<'a, 'b> {
+    frame: FramedMutSpace<'a, 'b>,
+}
+
+impl<'a, 'b> ByteWriter<'a, 'b> {
+    /// Create a new byte writer.
+    pub fn new(frame: FramedMutSpace<'a, 'b>) -> Self {
+        Self { frame }
+    }
+
+    /// Allocate memory in the atom, but don't write anything to it.
+    pub fn allocate(&mut self, size: usize) -> Option<&'a mut [u8]> {
+        self.frame.allocate(size, false).map(|(_, bytes)| bytes)
+    }
+
+    /// Copy data from the given slice to the atom.
+    pub fn write_raw(&mut self, bytes: &[u8]) -> Option<&'a mut [u8]> {
+        self.frame.write_raw(bytes, false)
+    }
+
+    /// Copy a struct instance to the atom.
+    pub fn write<T>(&mut self, instance: &T) -> Option<&'a mut T>
+    where
+        T: Unpin + Copy + Send + Sync + Sized + 'static,
+    {
+        (&mut self.frame as &mut dyn MutSpace).write(instance, false)
+    }
+}
+
+#[cfg(test)]
+#[cfg(feature = "host")]
+mod tests {
+    use crate::chunk::*;
+    use crate::*;
+    use std::mem::{size_of, size_of_val};
+    use urid::mapper::HashURIDMapper;
+    use urid::prelude::*;
+
+    #[test]
+    fn test_chunk_and_slice_writer() {
+        const SLICE_LENGTH: usize = 42;
+
+        let mapper = HashURIDMapper::new();
+        let map = Map::new(&mapper);
+        let urids = crate::AtomURIDCache::from_map(&map).unwrap();
+
+        let mut raw_space: Box<[u8]> = Box::new([0; 256]);
+
+        // writing
+        {
+            let mut space = RootMutSpace::new(raw_space.as_mut());
+            let frame = (&mut space as &mut dyn MutSpace)
+                .create_atom_frame(urids.chunk)
+                .unwrap();
+            let mut writer = Chunk::init(frame, ()).unwrap();
+
+            for (i, value) in writer
+                .allocate(SLICE_LENGTH - 1)
+                .unwrap()
+                .into_iter()
+                .enumerate()
+            {
+                *value = i as u8;
+            }
+            writer.write(&41u8).unwrap();
+        }
+
+        // verifying
+        {
+            let raw_space = unsafe {
```

Comment:
Right! This a leftover from when I used a `Box<u64>` to assure alignment. However, this made things more complicated and I switched to `Box<u8>`. Since this code works for both types, I missed it.